File: 2023/19_Aplenty/rule.py

# ======================================================================
# Aplenty
#   Advent of Code 2023 Day 19 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                             r u l e . p y
# ======================================================================
"Rule for the Advent of Code 2023 Day 19 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
COLON = ":"
LESS = "<"
MORE = ">"
CRITERIA = frozenset("xmas")
NOTEST = ""
NOBRANCH = None

# ======================================================================
#                                                                   Rule
# ======================================================================


class Rule(object):   # pylint: disable=R0902, R0903, R0205
    "Object for Aplenty"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.category = ""
        self.test = NOTEST
        self.value = 0
        self.branch = ""

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            self._process_text(text)

    def _process_text(self, text):
        "Assign values from text"

        # 0. Precondition axioms
        assert text is not None and len(text) > 0

        # 1. If there is no colon, goto only
        if COLON not in text:
            self.branch = text
            return

        # 2. Split off the criterial and branch name
        parts = text.split(COLON)
        assert len(parts) == 2
        self.branch = parts[1]

        # 3. Get the category and number
        if LESS in parts[0]:
            parts = parts[0].split(LESS)
            assert len(parts) == 2
            self.test = LESS
        else:
            parts = parts[0].split(MORE)
            assert len(parts) == 2
            self.test = MORE

        # 4. Save the category and number
        self.category = parts[0]
        self.value = int(parts[1])

        # 5. Postcondition axioms
        assert self.category in CRITERIA
        assert 0 < self.value <= 999999
        assert len(self.branch) > 0

    def evaluate(self, part):
        "Return branch if rule matches or NOBRANCH if it doesn't"

        # 1. If there is no test, branch
        if self.test == NOTEST:
            return self.branch

        # 2. Get the required value of the part
        value = part.value(self.category)

        # 3. If the test succeeds, branch
        if self.test == LESS and value < self.value:
            return self.branch
        if self.test == MORE and value > self.value:
            return self.branch

        # 4. We are not branching
        return NOBRANCH

# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------


if __name__ == '__main__':
    pass

# ======================================================================
# end                          r u l e . p y                         end
# ======================================================================

File: 2023/19_Aplenty/rule.go
// ======================================================================
// Aplenty
//   Advent of Code 2023 Day 19 -- Eric Wastl -- https://adventofcode.com
//
// go implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                          r u l e . g o
// ======================================================================
// Rule for the Advent of Code 2023 Day 19 puzzle

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"fmt"
	"strings"
)

// ----------------------------------------------------------------------
//
//	constants
//
// ----------------------------------------------------------------------
const ()

// ======================================================================
//                                                                   Rule
// ======================================================================

// Rule ... Object for Aplenty
type Rule struct {
	Part2    bool
	Text     string
	Category string
	MinMax   *Range
	Where    string
}

// NewRule ... Create a Rule object
func NewRule(part2 bool, text string) (*Rule, error) {
	// Rule Object for Aplenty

	// 1. Set the initial values
	p := new(Rule)
	p.Part2 = part2
	p.Text = text

	// 2. Get the ratings
	if len(text) > 0 {
		err := p.processText(p.Text)
		if err != nil {
			return nil, err
		}
	} else {
		p.MinMax, _ = NewRange(part2, "")
	}

	// 3. Return success
	return p, nil
}

// processText ... Assign values from text
func (p *Rule) processText(text string) error {

	// 1. Split the rule into conditional and target
	parts := strings.Split(text, ":")
	if len(parts) != 2 {
		return fmt.Errorf("invalid rule (colon) '%s'", text)
	}

	// 2. Save the target and the category
	p.Where = parts[1]
	p.Category = string(parts[0][0])

	// 3. Process the Range
	conditional := parts[0][1:]
	minmax, err := NewRange(p.Part2, conditional)
	if err != nil {
		return fmt.Errorf("invalid conditional '%s' in '%s'", conditional, text)
	}
	p.MinMax = minmax

	// 5. Return success
	return nil
}

// CheckPart ... Check a part against the rule, returning next rule or ""
func (p *Rule) CheckPart(part *Part) string {

	// 1. Get the value from the part
	value := part.GetValue(p.Category)

	// 2. If the value matches, the range go to the specified rule
	if p.MinMax.CheckValue(value) {
		return p.Where
	}

	// 3. No match, onto the next rule
	return ""
}

// String ... Representation of the rule
func (p *Rule) String() string {
	return p.Text
}

// ======================================================================
// end                          r u l e . g o                         end
// ======================================================================
